activerecord/test/cases/tasks/database_tasks_test.rb

```diff
@@ -467,6 +467,17 @@ class DatabaseTasksStructureLoadTest < ActiveRecord::TestCase
   end
 
   class DatabaseTasksCheckSchemaFileTest < ActiveRecord::TestCase
+    def setup
+      unless Rails.respond_to?(:root)
+        @defined_root = true
+        def Rails.root; :defined_root; end
```

Comment:
I don't think `redefine_method` can make this any simpler, since we still have to track whether `Rails.root` was defined beforehand to avoid removing it during teardown:

```patch
diff --git a/activerecord/test/cases/tasks/database_tasks_test.rb b/activerecord/test/cases/tasks/database_tasks_test.rb
index 9086a7c5b0..db5c3d4667 100644
--- a/activerecord/test/cases/tasks/database_tasks_test.rb
+++ b/activerecord/test/cases/tasks/database_tasks_test.rb
@@ -470,12 +470,12 @@ class DatabaseTasksCheckSchemaFileTest < ActiveRecord::TestCase
     def setup
       unless Rails.respond_to?(:root)
         @defined_root = true
-        def Rails.root; :defined_root; end
+        Rails.redefine_method(:root) { }
       end
     end

     def teardown
-      Rails.instance_eval { undef :root } if @defined_root
+      Rails.remove_possible_method(:root) if @defined_root
     end

     def test_check_schema_file
```